spec/syntax.md

```diff
@@ -905,24 +926,64 @@ backslash       = %x5C ; U+005C REVERSE SOLIDUS "\"
 
 ### Whitespace
 
-**_<dfn>Whitespace</dfn>_** is defined as one or more of
-U+0009 CHARACTER TABULATION (tab), 
-U+000A LINE FEED (new line),
-U+000D CARRIAGE RETURN, 
-U+3000 IDEOGRAPHIC SPACE, 
-or U+0020 SPACE.
+The syntax limits whitespace characters outside of a _pattern_ to the following:
+`U+0009 CHARACTER TABULATION` (tab), 
+`U+000A LINE FEED` (new line),
+`U+000D CARRIAGE RETURN`, 
+`U+3000 IDEOGRAPHIC SPACE`, 
+or `U+0020 SPACE`.
 
 Inside _patterns_ and _quoted literals_,
 whitespace is part of the content and is recorded and stored verbatim.
 Whitespace is not significant outside translatable text, except where required by the syntax.
 
+There are two whitespace productions in the syntax.
+**_<dfn>Optional whitespace</dfn>_** is whitespace that is not required by the syntax, 
+but which users might want to include to increase the readability of a _message_.
+**_<dfn>Required whitespace</dfn>_** is whitespace that is required by the syntax.
+
+Tools or users SHOULD insert paired isolating controls `U+2066 LEFT-TO-RIGHT ISOLATE`
+and `U+2069 POP DIRECTIONAL ISOLATE` as permitted by the ABNF _inside_
+of _placeholder_ markers `{` and `}` 
+and _outside_ _quoted-pattern_ markers `{{` and `}}`,
+as well as around _identifiers_,
+_literals_ (taking care not to include the mark inside any quotes), 
+or _option_ values that use right-to-left characters 
+so that the _message_ displays intelligibly.
+_Literals_, especially each individual _key_ in a variant, SHOULD be isolated 
+using these paired controls.
+Tools or users MAY also insert `U+200E LEFT-TO-RIGHT MARK` or `U+200F RIGHT-TO-LEFT MARK`
+characters where permitted by the syntax before or following _identifiers_,
+_literals_ (taking care not to include the mark inside any quotes), 
+or _option_ values that use right-to-left characters 
+so that the _message_ displays intelligibly.
+
+These definitions of _whitespace_ implement
+[UAX#31 Requirement R3a-2](https://www.unicode.org/reports/tr31/#R3a-2).
+It is a profile of R3a-1 in that specification because:
+the following pattern whitespace characters are not allowed:
+`U+000B FORM FEED`, 
+`U+000C VERTICAL TABULATION`, 
+`U+0085 NEXT LINE`, 
+`U+2028 LINE SEPARATOR` and 
+`U+2029 PARAGRAPH SEPARATOR`;
+the character `U+3000 IDEOGRAPHIC SPACE`
+_is_ interpreted as whitespace.
```

Comment:
```suggestion
the character `U+3000 IDEOGRAPHIC SPACE`
_is_ interpreted as whitespace,
as well as the directional isolates U+2066..U+2069
(these having the Default_Ignorable_Code_Point property, they are ignorable format controls).
```

Reply:
Is that true, though? I added them directly to `owsp`, meaning they can be used by themselves in optional whitespace locations (without having to insert any actual whitespace), while `wsp` requires that these characters (optionally) appear outside (before/after) a required run of whitespace. We don't impute any "whitespace-ness" to the controls. They're just doing their default-ignorable jobs 😉

Reply:
Yes, but being an ignorable format control (which does not count as a required space, and is only allowed as an optional space) counts as one of the three ways in which something can be _interpreted as whitespace_ in the sense of [UAX31-R3a](https://www.unicode.org/reports/tr31/#R3a) (the other two being horizontal space, and line terminator or part of a line terminator).

Reply:
(Yes, this is terribly confusing, and part of the general pattern that in Unicode contexts _whitespace_ means a host of different things, from [characters that are whitespace](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Awhitespace%3A%5D&g=&i=) to [characters that should be interpreted as whitespace in parsing](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Apattern_white_space%3A%5D&g=&i=) to [characters that are categorized as spaces](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Aspace+separator%3A%5D&g=&i=) to the zero width space and no-break space that are called spaces, but are not spaces at all and instead are controls serving as overrides for some segmentation algorithms.

And then of course you have [characters that behave like space in line breaking](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Aline_break%3Dspace%3A%5D&g=&i=), [characters that behave like space in bidirectional ordering](https://util.unicode.org/UnicodeJsps/list-unicodeset.jsp?a=%5B%3Abc%3Dwhite_space%3A%5D&g=&i=), etc.

Unicode Properties & Algorithms are an interesting place.)

Reply:
Yes, I'm aware. Note well that these controls are not in the `s` production, which is what we call whitespace. I realize that other interpretations of whitespace are possible in Unicode. Your note addition would, I think, be confusing, since the controls are optional and are definitely *not* counted as whitespace in required whitespace parts of our syntax.

Is there harm in leaving this out? If so, I'll commit it.

Reply:
I think it the conformance statement really ought to mention that you are adding to the set of ignorable format controls, but I agree that my proposed wording is needlessly pedantic and confusing. How about this:
```suggestion
the character `U+3000 IDEOGRAPHIC SPACE`
_is_ interpreted as whitespace,
 and the directional isolates U+2066..U+2069
 are treated as ignorable format controls.
```

Reply:
Works for me! Done.